### ISSUE TITLE
The addon does not work correctly in blender 4.3.0 (+4.3.1)

### ISSUE THREAD

**Issue by 8Yaron8** [shteeve3d/blender-wiggle-2#87]
When loading an old project into a new version of blender, the characters' clothes and hair moved unnaturally jerkily and changed in size, mostly increasing. Resetting to the original position does not work. If you try to use the addon on a new model, it looks like everything works fine, until you add collision function.

**Comment by weThere** [comment]
I do not know if the following work as I've not used wiggle2 since blender 4.x.

You can attempt to modify your addon script manually, reference for blender 4.x compatibility:               This might fix it (and other 4.2+ incompatibilities): https://blenderartists.org/t/wiggle-2/1454788/206

_Originally posted by @KaijuKoder2 in https://github.com/shteeve3d/blender-wiggle-2/issues/84#issuecomment-2455912794_

and this potentially for solving collision issues:
@LuciferTC9527
https://github.com/shteeve3d/blender-wiggle-2/issues/68#issuecomment-2267100660

If you're in a similar position to me and at the moment have other priorities than scripting, wait and see if someone else takes over updating this addon, use blender 3.x for wiggle2 projects or find other free/paid intermediary addons.

The most "recent" version of this addon is by @schroef, adding in physics presets, I do not know if @schroef as well updated blender compatibility with this: https://github.com/schroef/blender-wiggle-2


(Irritatingly I've yet to find a physics addon that handles bone head physics, e.g., allowing physics on IK controls, there is always the cheapest freest option of doing some cheeky workaround for cyclical bone constraints (most notable is the damped track chain) directly in blender to get physics effects as well.).)

**Comment by 8Yaron8** [comment]
I see, I tried to replace the code that you provided in the link, but it also doesn't work, and I can't fix the addon myself because I don't know how to write scripts. The addon still works in blender 4.2.4, so I'll use it for now, but I haven't baked any animations yet.

**Comment by schroef** [comment]
I have seen issues in bl4.1.1, I also didn't have to update code to get it to work in this version.

I did see something happening indeed when millions is used and the bones are 2 close. There is not a distance parameter (yet)

@8Yaron8 
Ps did you reset the bone or did you use reset physics in the wiggle bone 2 panel?
Probably physics is the issue

**Comment by 8Yaron8** [comment]
Yes, I tried to do this too, but nothing works, only cancellation (CTRL+Z) works and physics with bone size go crazy again.

**Comment by KaijuKoder2** [comment]
I'm still using 4.1, but Wiggle 2 author @shteeve3d says here that it does work on 4.3: https://blenderartists.org/t/wiggle-2/1454788/217

**Comment by 8Yaron8** [comment]
Yes, but it works partially if you do not use the collision functions. I used the addon to the models in the previous version of blender and it works stably there, but if you load the project in the new version, literally Harlem Shake begins. Later I tried to add physics in the new version first, everything seems to work fine until it comes to collision. In that forum in the last messages another person has the same problem.

**Comment by 3dcinetv** [comment]
Blender 4.3 is a solid release now. Would it be possible to have an official update for Wiggle 2, please?
If I activate the tail/head for wiggle bone, the head control in a rigify rig doesn't even move (scene is too heavy when wiggle 2 is active in 4.3). Even if I solo the armature, playback it's at 3fps on a 52 GB ram with RTX3060.
![image](https://github.com/user-attachments/assets/5b675b25-3df4-40a6-949b-2bae18ed6f6d)

**Comment by schroef** [comment]
I doubt video card has anything to do with this. I think this runs on cpu.

Does 4.3 have updates in the post and pre handlers perhaps?

**Comment by Hans-xwh** [comment]
I noticed that the bug where the bones go all crazy is related to physics, but sometimes even when you disable the physics trough the addon panel, they'll still be enabled, even when the collision appear as not enabled.

To fix this, go to the bone custom properties, find the collider object or collection parameter, and manually remove the reference. You can also select all the enabled bones, and when clicking the remove reference button hold alt, and they will be fixed altogether.

**Comment by KDuffs** [comment]
A good way to tell something isn't working is when the 'Reset Physics' button does nothing.

Removing all collisions brings back 'Reset Physics'. 

Using the same file:

In 4.2.X, the collisions work as intended. All normals are correct.

In 4.3.X, the collisions cause the bones to react as if normals are inside out, they jitter and bug out on what seems to be a compounding rate.

With 4.4.X (alpha builds) the collision simply isn't registered- as if it hasn't been added. Reset Physics still doesn't work however until the collision property has been removed from the bones.

**Comment by KDuffs** [comment]
FINALLY FIXED!*

Collisions appear to be correct in @Labhatorian's incredible fork using 4.3. Brilliant!!

*pre-roll baking still seems to be at issue

**Comment by 8Yaron8** [comment]
> FINALLY FIXED!
> 
> Collisions appear to be correct in [@Labhatorian](https://github.com/Labhatorian)'s incredible fork using 4.3. Brilliant!!

Just checked and it works on the model that I have checked many times, also works with the previous version 4.2.6. Only the loop  + pre-roll after baking does not work correctly, the bones fly up and return to their place.

**Comment by KDuffs** [comment]
> > FINALLY FIXED!
> > Collisions appear to be correct in [@Labhatorian](https://github.com/Labhatorian)'s incredible fork using 4.3. Brilliant!!
> 
> Just checked and it works on the model that I have checked many times, also works with the previous version 4.2.6. Only the loop + pre-roll after baking does not work correctly, the bones fly up and return to their place.

Yes I since tried baking and found this to be the case. Looping seems to work for me if I don't bake it?

As for baking itself, the 'live physics' seems to be rendering now without having to bake, so that could be a workaround?

**Comment by 8Yaron8** [comment]
In my case I have a big scene and 3 characters that use physical bones, with bone physics enabled the preview slows down to 10-5 frames per second.

**Comment by KDuffs** [comment]
> In my case I have a big scene and 3 characters that use physical bones, with bone physics enabled the preview slows down to 10-5 frames per second.

yes we could definitely do with the baking option to be fixed.

**Comment by 3dcinetv** [comment]
Baking wiggle in Blender 4.3 is necessary to make the addon useful again, baking prevents the weird stretching because of physics (0 frame start) and reset. Whatever the case, I hope we can finally solve this issue before Blender 4.4 which also inserts new API problems for baking (multi-channels).

**Comment by LuciferTC9527** [comment]
If you are using the @Labhatorian branch and there is no physical parameter modification option you can replace the following function

![Image](https://github.com/user-attachments/assets/bb252d29-5286-45b3-ae55-3fbcbffdb8b5)

![Image](https://github.com/user-attachments/assets/e569009f-f425-4f39-9d6d-f5578d27e93b)

![Image](https://github.com/user-attachments/assets/02893ce6-dd2a-4a0b-b47d-cd3a80b3d64f)

![Image](https://github.com/user-attachments/assets/dd6953b6-9f7b-4ce7-9bce-fcecfda146ad)

```
def set_props_for_selected(attr, value):
    for bone in selected_bones:
        try:
            setattr(bone, attr, value)
        except AttributeError:
            pass
```

![Image](https://github.com/user-attachments/assets/e4466230-fa01-4f0f-bcca-4d14170e8bce)

**Comment by 3dcinetv** [comment]
can you share the updated .py to download from somewhere, please? I'm not a coder. Thanks.

**Comment by LuciferTC9527** [comment]
There will be some problems in the first frame after baking, I don't know if it's the branch or my problem
https://drive.google.com/file/d/1F4zVWB6UkjALyOuPyJRXp3i5-619_SdX/view?usp=drive_link

**Comment by ItsMeNicoBros** [comment]
> There will be some problems in the first frame after baking, I don't know if it's the branch or my problem https://drive.google.com/file/d/1F4zVWB6UkjALyOuPyJRXp3i5-619_SdX/view?usp=drive_link

It does wonders already, but sadly it has a problem where I can't use the collision meshes. When I use them, the physics simply deactivate themselves. That's kinda sad, since most of my works do has some heavy usage of collision meshes to keep everything on proper places.

It's great. but it's still not enough to migrate from 4.2 to anything above it, since this is mostly for some basic wiggle stuff 😭

**Comment by schroef** [comment]
> > There will be some problems in the first frame after baking, I don't know if it's the branch or my problem https://drive.google.com/file/d/1F4zVWB6UkjALyOuPyJRXp3i5-619_SdX/view?usp=drive_link
> 
> It does wonders already, but sadly it has a problem where I can't use the collision meshes. When I use them, the physics simply deactivate themselves. That's kinda sad, since most of my works do has some heavy usage of collision meshes to keep everything on proper places.
> 
> It's great. but it's still not enough to migrate from 4.2 to anything above it, since this is mostly for some basic wiggle stuff 😭 

Did you try the other version from @labhatorian

https://github.com/Labhatorian/blender-wiggle-2

**Comment by 3dcinetv** [comment]
Yes, I've tried Labhatorian's wiggle setup. It works in the viewport and all, but crashes my .blend when I try to bake. 
I used to think I had too many "addons active". But that wasn't the issue. Dynamic "reset"/"rest"/"frame behind evaluation" are all part of python's erratic way to work with the Blender viewport. I am still using Blender 4.3. Everything else (including pushing NLA clips) feels so erratic because of the new changes in the API for B44, 45. Incidentally, not even Claude can repair .py wiggle bones.
